ci/codecov.sh

```diff
@@ -23,51 +23,69 @@ set -ex
     export B2_VARIANT=debug
     if [ -z "$B2_CI_VERSION" ]; then
-        export B2_CXXFLAGS="${B2_CXXFLAGS:+$B2_CXXFLAGS }cxxflags=-fkeep-static-functions cxxflags=--coverage"
+        export B2_CXXFLAGS="${B2_CXXFLAGS:+$B2_CXXFLAGS }cxxflags=-fprofile-arcs cxxflags=-ftest-coverage cxxflags=--coverage"
     else
-        export B2_CXXFLAGS="${B2_CXXFLAGS:+$B2_CXXFLAGS }-fkeep-static-functions --coverage"
+        export B2_CXXFLAGS="${B2_CXXFLAGS:+$B2_CXXFLAGS }-fprofile-arcs -ftest-coverage --coverage"
     fi
     export B2_LINKFLAGS="${B2_LINKFLAGS:+$B2_LINKFLAGS }--coverage"
 
 elif [[ "$1" == "upload" ]]; then
     if [ -z "$GCOV" ]; then
-        ver=7 # default
-        if [ "${B2_TOOLSET%%-*}" == "gcc" ]; then
-            if [[ "$B2_TOOLSET" =~ gcc- ]]; then
-                ver="${B2_TOOLSET##*gcc-}"
-            elif [[ "$B2_COMPILER" =~ gcc- ]]; then
-                ver="${B2_COMPILER##*gcc-}"
+        # enforce.sh ensures B2_COMPILER
+        COMPILER=${B2_COMPILER}
+        COMPILER_FAMILY=$(echo ${COMPILER} | cut -d'-' -f1)
+        COMPILER_EDITION=$(echo ${COMPILER} | cut -d'-' -f2)
+        # clang is still not working - need a llvm-cov wrapper and fixup pathing
+        if [ "${COMPILER_FAMILY}" == "clang" ]; then
+            echo "clang is not supported for coverage yet"
+            exit 1
+            # find /usr/lib -name 'llvm-cov*' -print
+            # LLVM_COV_PATH=$(find /usr/lib -type f -name llvm-cov -exec dirname {} \;)
+            # export PATH=${LLVM_COV_PATH}:${PATH}
+            # GCOV=llvm-cov
+        elif [ "${COMPILER_FAMILY}" == "gcc" ]; then
+            # cutting "gcc" above yields "gcc" for both values if no version present
+            if [ "${COMPILER_EDITION}" == "${COMPILER_FAMILY}" ]; then
+                GCOV=gcov
+            else
+                GCOV=gcov-${COMPILER_EDITION}
             fi
+        else
+            echo "Cannot determine GCOV."
+            exit 1
         fi
-        GCOV=gcov-${ver}
     fi
 
     # install the latest lcov we know works
     rm -rf /tmp/lcov
     cd /tmp
-    git clone --depth 1 -b v1.14 https://github.com/linux-test-project/lcov.git
+    git clone --depth 1 -b v1.15 https://github.com/linux-test-project/lcov.git
     export PATH=/tmp/lcov/bin:$PATH
     command -v lcov
     lcov --version
 
     # switch back to the original source code directory
     cd $BOOST_CI_SRC_FOLDER
-    : "${LCOV_BRANCH_COVERAGE:=1}" # Set default
+    : "${LCOV_BRANCH_COVERAGE:=1}" # On by default, job can override but advise against it
 
     # coverage files are in ../../b2 from this location
-    lcov --gcov-tool=$GCOV --rc lcov_branch_coverage=${LCOV_BRANCH_COVERAGE} --base-directory "$BOOST_ROOT/libs/$SELF" --directory "$BOOST_ROOT" --capture --output-file all.info
-    # dump a summary on the console
-    lcov --list all.info
+    lcov --rc lcov_branch_coverage=${LCOV_BRANCH_COVERAGE} --gcov-tool=${GCOV} --directory "${BOOST_ROOT}" --capture --output-file all.info
```

Comment:
Why can the `--base-directory "$BOOST_ROOT/libs/$SELF"` param be removed? It doesn't seem to be required in your test run but I don't really understand the `--directory` / `--base-directory` meanings and am reluctant to change that, as it works now.

Reply:
@Flamefire this is why: https://github.com/boostorg/uuid/runs/5060707720?check_suite_focus=true#step:13:99

Without the base directory and other changes I made, it works.

Reply:
Here's with LCOV 1.15 and without base-directory:

https://github.com/boostorg/uuid/runs/5060934544?check_suite_focus=true#step:13:99

```
+ lcov --rc lcov_branch_coverage=1 --gcov-tool=gcov-10 --directory /home/runner/work/uuid/boost-root --capture --output-file all.info
Capturing coverage data from /home/runner/work/uuid/boost-root
Found gcov version: 10.3.0
Using intermediate gcov format
Scanning /home/runner/work/uuid/boost-root for .gcda files ...
Found [108](https://github.com/boostorg/uuid/runs/5060934544?check_suite_focus=true#step:13:108) data files in /home/runner/work/uuid/boost-root
Processing chrono/build/gcc-10/debug/cxxstd-11-iso/link-static/threading-multi/visibility-hidden/chrono.gcda
Processing chrono/build/gcc-10/debug/cxxstd-03-iso/link-static/threading-multi/visibility-hidden/chrono.gcda
Processing chrono/build/gcc-10/debug/address-model-32/cxxstd-11-iso/link-static/threading-multi/visibility-hidden/chrono.gcda
Processing chrono/build/gcc-10/debug/address-model-32/cxxstd-03-iso/link-static/threading-multi/visibility-hidden/chrono.gcda
Processing timer/build/gcc-10/debug/cxxstd-11-iso/link-static/threading-multi/visibility-hidden/cpu_timer.gcda
Processing timer/build/gcc-10/debug/cxxstd-11-iso/link-static/threading-multi/visibility-hidden/auto_timers_construction.gcda
Processing timer/build/gcc-10/debug/cxxstd-03-iso/link-static/threading-multi/visibility-hidden/cpu_timer.gcda
Processing timer/build/gcc-10/debug/cxxstd-03-iso/link-static/threading-multi/visibility-hidden/auto_timers_construction.gcda
Processing timer/build/gcc-10/debug/address-model-32/cxxstd-11-iso/link-static/threading-multi/visibility-hidden/cpu_timer.gcda
```